src/tools/crawler/network.rs

```diff
@@ -0,0 +1,101 @@
+use std::{
+    cmp::Ordering,
+    collections::{HashMap, HashSet},
+    hash::{Hash, Hasher},
+    net::SocketAddr,
+    time::Duration,
+};
+
+use tokio::{sync::RwLock, time::Instant};
+use tracing::{error, info};
+
+#[derive(Default)]
+pub struct KnownNetwork {
+    nodes: RwLock<HashMap<SocketAddr, KnownNode>>,
+    connections: RwLock<HashSet<KnownConnection>>,
+}
+
+impl KnownNetwork {
+    pub(super) async fn insert_node(&self, addr: SocketAddr) {
+        let mut nodes = self.nodes.write().await;
+        if !nodes.contains_key(&addr) {
+            nodes.insert(addr, KnownNode::default());
+            info!("Known nodes: {}", nodes.len());
+        }
```

Comment:
we might just use the [insert](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert) fn directly and use the return value to printout the info trace

Reply:
True, but the thing is I don't want to override the struct in the map as it contains some stats.

Reply:
right! then try_insert would be actually useful here, but a stable feature yet